S2 CardView and ghost loading

devongovett opened the pull request:
This adds S2 Cards, which come in the following configurations:

* `Card` - a generic card container, into which you can place a `<CardPreview>`, `<Content>` with `title`, `description` and action menu slots, and a `<Footer>`. Cards come in 4 sizes, 3 densities, and 4 variants. A `<CollectionCardPreview>` can also be used to display a grid of multiple images in a preview.
* `AssetCard` - a pre-configured `Card` for displaying an asset, with an image preview and metadata.
* `UserCard` - a pre-configured `Card` for displaying a user, with an optional preview, avatar, and metadata.
* `ProductCard` - a pre-configured `Card` for displaying a product, with an optional preview, thumbnail, metadata, and footer call to action.

`CardView` displays a grid or waterfall layout of cards, with selection, keyboard navigation, item actions/links, infinite loading, etc.

devongovett commented on packages/@react-spectrum/s2/src/Content.tsx:
```diff
@@ -102,17 +103,26 @@ export const TextContext = createContext<ContextValue<ContentProps, DOMRefValue>
     return null;
   }
+
+  slot = slot && racContext && 'slots' in racContext && !racContext.slots?.[slot] ? undefined : slot;
```
Not sure if we want this or not. The idea is to not forward Spectrum-specific slots that RAC doesn't expect

snowystinger replied:
Seems related: https://github.com/adobe/react-spectrum/issues/6992

devongovett commented on packages/@react-spectrum/s2/src/Form.tsx:
```diff
@@ -36,11 +37,29 @@ export interface FormProps extends FormStyleProps, Omit<RACFormProps, 'className
+    }
+
+    return result;
+  }, [ctx, props, isSkeleton]);
```
Is it a bad idea to do this here? It was an easy way to mark all form components within a skeleton as disabled.

devongovett replied:
So, the alternative is to manually do this in every individual component rather than here in one place. Or I guess we could rename `useFormProps` to something else...

devongovett commented on packages/@react-stately/virtualizer/src/LayoutInfo.ts:
```diff
@@ -36,6 +36,11 @@ export class LayoutInfo {
+  /** 
+   * Content for this view if it was generated by the layout rather than coming from the Collection.
+   */
+  content: any | null;
```
Ideally we'd use a generic here but that would require a lot of changes to propagate everywhere...

snowystinger replied:
lot of changes, but would they be breaking?

devongovett commented on packages/@react-spectrum/s2/src/CardView.tsx:
```diff
@@ -0,0 +1,511 @@
+  onLoadMore?: () => void
+}
+
+class FlexibleGridLayout<T extends object, O> extends Layout<Node<T>, O> {
```
This is different from the GridLayout in RAC because it supports variable row heights - the height of a row is equal to the maximum height of the items in that row. The RAC GridLayout just scales the sizes of its items proportionally, which won't work for wrapping text. I could see potentially adding this as an option in RAC at some point.

snowystinger commented on packages/@react-spectrum/s2/src/ImageCoordinator.tsx:
```diff
@@ -0,0 +1,150 @@
+}
+
+export function ImageCoordinator(props: ImageCoordinatorProps) {
+  // If we are already inside another ImageList, just pass
```
```suggestion
  // If we are already inside another ImageCoordinator, just pass
```
Is this what it's saying? Or can there be multiple lists in one coordinator?

devongovett replied:
Nested ImageCoordinators don't do anything, they just delegate to the parent.

snowystinger commented on packages/@react-spectrum/s2/src/Link.tsx:
```diff
@@ -83,15 +85,27 @@ const link = style<LinkRenderProps & LinkStyleProps>({
+
+  useLayoutEffect(() => {
+    if (domRef.current) {
+      // TODO: should RAC Link pass through inert?
```
I think so, seems like an ok thing to pass through in general and might make it easier for people to make their website/form builders instead of having to wrap all their components

snowystinger commented on packages/@react-spectrum/s2/src/SkeletonCollection.tsx:
```diff
@@ -0,0 +1,36 @@
+export const SkeletonCollection = createLeafComponent('skeleton', (props: SkeletonCollectionProps, ref, node) => {
+  // Cache rendering based on node object identity. This allows the children function to randomize
+  // its content (e.g. heights) and preserve on re-renders.
+  // TODO: do we need a `dependencies` prop here?
```
Can you expand on why we might need dependencies here? what kinds of things can change where we'd want to update?

devongovett replied:
just like `<Collection>`, this component caches the results. So if the child function depends on values from outside, they'd need to be dependencies so that the cache invalidates.

snowystinger commented on packages/@react-stately/virtualizer/src/Virtualizer.ts:
```diff
@@ -114,8 +114,8 @@ export class Virtualizer<T extends object, V> {
   private _renderView(reusableView: ReusableView<T, V>) {
-    let {type, key} = reusableView.layoutInfo;
-    reusableView.content = this.collection.getItem(key);
+    let {type, key, content} = reusableView.layoutInfo;
```
I don't quite follow the implications of this change

devongovett replied:
yep. this enables the layout to produce items that aren't in the original collection.

snowystinger commented on packages/dev/parcel-transformer-s2-icon/IconTransformer.js:
```diff
@@ -57,9 +57,10 @@ module.exports = new Transformer({
         plugins: ['@svgr/plugin-svgo', '@svgr/plugin-jsx']
       })
-    ).replace('export default SvgComponent;', '');
+    ).replace('export default ForwardRef;', '');
```
can delete comment below?

devongovett commented on packages/@react-aria/utils/src/useLoadMore.ts:
```diff
@@ -29,7 +29,7 @@ export interface LoadMoreProps {
   scrollOffset?: number,
   /** The data currently loaded. */
-  items?: any[]
+  items?: any
```
Iterable in this case but we don't actually use the value, it's just for invalidating an effect.

devongovett commented on packages/@react-spectrum/s2/src/ActionButton.tsx:
```diff
@@ -175,6 +177,7 @@ export const ActionButtonContext = createContext<ContextValue<ActionButtonProps,
 
 function ActionButton(props: ActionButtonProps, ref: FocusableRef<HTMLButtonElement>) {
   [props, ref] = useSpectrumContextProps(props, ref, ActionButtonContext);
+  props = useFormProps(props as any);
```
Disabled

snowystinger commented on packages/@react-spectrum/s2/src/Card.tsx:
```diff
@@ -0,0 +1,669 @@
+  let {UNSAFE_className, UNSAFE_style} = props;
+  let domRef = useDOMRef(ref);
+  return (
+    <div
```
should we add filterDOMProps to all of these, people always want to pass their data-*s

snowystinger commented on packages/@react-spectrum/s2/src/Card.tsx:
```diff
@@ -0,0 +1,669 @@
+    ':nth-last-child(4):first-child': '[3/2]'
+  },
+  gridColumnEnd: {
+    ':nth-last-child(4):first-child': 'span 3'
```
hah, nice selector, good way to only do it for first image in a set of 4

snowystinger commented on packages/@react-spectrum/s2/src/Card.tsx:
```diff
@@ -0,0 +1,669 @@
+          [AvatarContext, {
+            size: avatarSize[size],
+            UNSAFE_style: {
+              '--size': avatarSize[size] + 'px'
```
we've got to start an audit of all these variables we introduce

or hash them for internal module use or something. it'd be nice if we could export/import them internally as well, so we know where they are defined vs used

don't worry about in this PR

snowystinger commented on packages/@react-spectrum/s2/src/CardView.tsx:
```diff
@@ -0,0 +1,513 @@
+    return false;
+  }
+
+  // Override keyboard navigation to work spacially.
```
Technically these overrides work for all grid layouts if you include keyAbove/keyBelow the same way. It's just that in a grid there's only one overlap candidate ever possible.

Any reason not to change how the base version works to this and then not worry about overriding?

devongovett replied:
The spacial navigation algorithm isn't that simple to implement performantly with just the DOM. There's not a good way to query for elements inside a rectangle that I know of

snowystinger commented on packages/@react-spectrum/s2/src/CardView.tsx:
```diff
@@ -0,0 +1,513 @@
+    default: 'auto',
+    isLoading: 'hidden'
+  },
+  scrollPadding: '[18px]',
```
i know @LFDanLu was having some issues with inconsistency across browsers, did you resolve that?

devongovett replied:
actually this one isn't used. it gets overridden below by the inline style. removing this.

snowystinger commented on packages/@react-spectrum/s2/src/CardView.tsx:
```diff
@@ -0,0 +1,513 @@
+  let {children, layout: layoutName = 'grid', size = 'M', density = 'regular', variant = 'primary', selectionStyle = 'checkbox', UNSAFE_className = '', UNSAFE_style, styles, ...otherProps} = props;
+  let options = layoutOptions[size][density];
+  let layout = useMemo(() => {
+    variant; // needed to invalidate useMemo
```
i think this is preferable over eslint-ignoring the entire dep array, should update this in other places. outside scope of pr

snowystinger commented on packages/@react-spectrum/s2/src/CardView.tsx:
```diff
@@ -0,0 +1,513 @@
+  let ref = useRef(null);
+  useLoadMore({
+    isLoading: props.loadingState !== 'idle' && props.loadingState !== 'error',
+    items: props.items, // TODO: ideally this would be the collection. items won't exist for static collections, or those using <Collection>
```
this is fine though because a static collection wouldn't ever be in a loadMore/loading state?

devongovett replied:
If you used `<Collection>` or another component that produced multiple children it could.

snowystinger replied:
ahhh ok, now i understand the comment better.
so the issue here is that we haven't built the collection yet, that happens below.

devongovett replied:
yeah

snowystinger reviewed (commented):
Noticed a bit of a bug with virtualizer reuse. Selection can appear to move to other un-selected items when moving around via keyboard. I know we have this in TableView and ListView, but it feels more noticeable because of the more obvious selected border.

https://github.com/user-attachments/assets/69aea1f0-84de-477b-9a9c-dec8f423eb32

The docs page for Cards has a warning about missing labels.

devongovett commented on packages/@react-spectrum/s2/src/Skeleton.tsx:
```diff
@@ -0,0 +1,142 @@
+      );
+      animationRef.current.startTime = 0;
+    } else if (!isAnimating && animationRef.current) {
+      animationRef.current.cancel();
```
Animations don't continue after unmounting

devongovett commented on packages/@react-spectrum/s2/src/Skeleton.tsx:
```diff
@@ -0,0 +1,142 @@
+  * {
+    visibility: hidden;
+  }
+`, 'UNSAFE_overrides');
```
Maybe but I don't want to refactor the entire style system just to support cards haha

snowystinger replied:
sure, we can follow it up if it presents an issue

devongovett commented on packages/@react-spectrum/s2/src/Card.tsx:
```diff
@@ -0,0 +1,669 @@
+  alignItems: 'end',
+  justifyContent: 'space-between',
+  gap: 8,
+  paddingTop: '[calc(var(--card-spacing) * 1.5 / 2)]'
```
It's spectrum design math from a document Lindsay made

snowystinger replied:
haha, ok, then i guess keep as is so it matches design

reidbarber commented on packages/@react-spectrum/s2/src/Card.tsx:
```diff
@@ -0,0 +1,669 @@
+  [props] = useSpectrumContextProps(props, ref, CardContext);
+  let domRef = useDOMRef(ref);
+  let {density = 'regular', size = 'M', variant = 'primary', UNSAFE_className = '', UNSAFE_style, styles} = props;
+  let isQuiet = variant === 'quiet';
```
<img width="579" alt="Screenshot 2024-09-03 at 2 28 56 PM" src="https://github.com/user-attachments/assets/77db3c3f-d269-48e5-8760-cafd8cf6fbea">

Should the background be different for the quiet variant here, or are we using the wrong background in storybook?

snowystinger replied:
one of the non-supported combinations?

devongovett replied:
hmm, yes, the background is wrong in storybook docs. I originally had `transparent` here but then realized I shouldn't override `--s2-container-bg` in that case (it is used for the border around avatars etc.). Guess I should add it back.

reidbarber commented on packages/@react-spectrum/s2/stories/CardView.stories.tsx:
```diff
@@ -0,0 +1,231 @@
+    <Card href={topic.links.html} target="_blank" textValue={topic.title}>
+      <CollectionCardPreview>
+        {topic.preview_photos.slice(0, 4).map(photo => (
+          <Image key={photo.id} alt="" src={photo.urls.small} />
```
Would we support wrapping these in links?

reidbarber reviewed (commented):
The styles and behavior look great! Just a few initial comments

LFDanLu commented on packages/@react-spectrum/s2/src/CardView.tsx:
```diff
@@ -0,0 +1,513 @@
+  outlineStyle: 'none'
+}, getAllowedOverrides({height: true}));
+
+export function CardView<T extends object>(props: CardViewProps<T>) {
```
I remember for the old CardView implementation that we offered the ability to customize the layout via maxColumns, inter row/column padding, etc. I assume these were scrapped by Spectrum but just wanna make sure.

LFDanLu commented on packages/@react-spectrum/s2/src/CardView.tsx:
```diff
@@ -0,0 +1,513 @@
+                ...UNSAFE_style,
+                scrollPadding: options.minSpace.height
+              }}
+              className={renderProps => UNSAFE_className + cardViewStyles({...renderProps, isLoading: props.loadingState === 'loading'}, styles)}>
```
Will need a focus ring when the CardView is empty and tabbed to.

LFDanLu reviewed (commented):
Still going through the Skeleton and layout logic, some initial comments from testing. Overall, performs nicely, tested a variety of keyboard operations/prefers reduced motion/etc

devongovett commented on packages/@react-spectrum/s2/src/Skeleton.tsx:
```diff
@@ -0,0 +1,142 @@
+  let childRef = 'ref' in children ? children.ref as any : children.props.ref;
+  return (
+    <SkeletonContext.Provider value={null}>
+      {isLoading ? cloneElement(children, {
```
No, `SkeletonWrapper` is not exported publicly.

devongovett commented on packages/@react-spectrum/s2/src/ImageCoordinator.tsx:
```diff
@@ -0,0 +1,150 @@
+
+  useEffect(() => {
+    if (!loadedAll) {
+      let timeoutId = setTimeout(() => {
```
how could there be multiple at the same time? The effect cleanup will run when `timeout` changes and will clear the previous one.

snowystinger replied:
ah yep, over thought it or missed that

devongovett commented on packages/@react-spectrum/s2/src/Image.tsx:
```diff
@@ -0,0 +1,207 @@
+    // If the image is already loaded, update state immediately instead of waiting for onLoad.
+    if (state === 'loading' && imgRef.current?.complete) {
+      // Queue a microtask so we don't hit React's update limit.
+      // TODO: is this necessary?
```
It's sort of hard to tell if I'm seeing placebo or how much difference it actually makes, but I noticed that the image's `complete` property could be true before the `onLoad` event fires (async). So the theory here is that we could show the image slightly earlier, which is important while scrolling. Just not sure how much difference it makes.

devongovett commented on packages/@react-spectrum/s2/stories/CardView.stories.tsx:
```diff
@@ -0,0 +1,231 @@
+            width: 'full',
+            pointerEvents: 'none'
+          })}
+          UNSAFE_style={{
```
not the theme. These values are dynamic: `${item.width} / ${item.height}`. I did wonder if we should somehow support "safe" inline styles (limited overridable properties) for cases like this. Probably a larger discussion than cards though.

snowystinger replied:
yeah, I think some safe ones should be provided for places where we know it to be so

devongovett commented on packages/@react-spectrum/s2/stories/CardView.stories.tsx:
```diff
@@ -0,0 +1,231 @@
+    <CardView 
+      aria-label="Assets"
+      {...args}
+      UNSAFE_style={{height: viewMode === 'docs' ? 600 : '100vh', width: viewMode === 'docs' ? '100%' : '100vw'}}
```
well.... those values aren't in the theme. Sizes only go up to 384px, and we don't have `100vw`.

devongovett commented on packages/@react-spectrum/s2/src/Card.tsx:
```diff
@@ -0,0 +1,669 @@
+    forcedColors: '[0 0 0 1px ButtonBorder]',
+    variant: {
+      tertiary: {
+        // Render border with box-shadow to avoid affecting layout.
```
idk... I don't like this at all, and I don't want to encourage using box-shadow as a border. It doesn't show in high contrast mode by default and feels like a hack. Ideally `outline` would be used, but here we already used outline for the focus ring.

devongovett commented on packages/@react-spectrum/s2/src/Card.tsx:
```diff
@@ -0,0 +1,669 @@
+    default: 0,
+    isSelected: 1
+  }
+  // outlineColor: 'white',
+  // outlineOffset: -4,
+  // outlineStyle: 'solid',
+  // outlineWidth: 2
```
waiting for design

devongovett commented on packages/@react-spectrum/s2/src/Card.tsx:
```diff
@@ -0,0 +1,669 @@
+  let press = pressScale(domRef, UNSAFE_style);
+  return (
+    <ElementType
+      {...props}
```
that happens inside GridListItem but probably needed in the div case above.

devongovett commented on packages/@react-spectrum/s2/stories/CardView.stories.tsx:
No, density at the CardView level only affects the spacing between the cards, not within the cards. Density at the card level affects the spacing within the cards (but there is no story control for that).

devongovett commented on packages/@react-spectrum/s2/stories/CardView.stories.tsx:
```diff
@@ -0,0 +1,231 @@
+  );
+}
+
+export const Example = (args: CardViewProps<any>, {viewMode}) => {
```
hmm I haven't been able to reproduce this. any idea how you got this?

reidbarber replied:
Not able to reproduce now

snowystinger reviewed (dismissed):
Testing was pretty good, happy to get in and address anything else in new PRs

snowystinger commented on packages/@react-aria/virtualizer/src/ScrollView.tsx:
```diff
@@ -177,27 +182,38 @@ export function useScrollView(props: ScrollViewProps, ref: RefObject<HTMLElement
+      // We also need to wait until all refs are set (e.g. when passing a ref down from a parent).
+      // If we are in an `act` environment, update immediately without a microtask so you don't need
+      // to mock timers in tests. In this case, the update is synchronous already.
+      if (globalThis.IS_REACT_ACT_ENVIRONMENT) {
```
possibly not going to work, as some Issues out there recommend setting this to false in the tests... 
https://github.com/testing-library/react-testing-library/issues/1108
I can't imagine this is common, but I imagine this will be fun for people to debug, hahaimgonnagocrynow

snowystinger reviewed (approved):
I still see the horizontal scrollbar on initial render. I'm not fussed about it. It goes away if I resize the page.

devongovett commented:
> I still see the horizontal scrollbar on initial render

That's embarrassing. I had it working and then optimized it and apparently it broke and I didn't notice? Reverted to the previous version I had, which updates more frequently (whenever the content size changes at all) rather than trying to detect only when the scrollbars show and hide.

devongovett commented:
Updated with the following changes based on my meeting with design today:

* Checkbox wrapper changed from 6px to 4px
* An extra inner white/black stroke was added to quiet cards without a checkbox in order to distinguish the selection outline from the preview image. All other card variants do not have this stroke.
* The whole card now has press scaling in all cases rather than only the checkbox. For quiet cards, only the preview scales down.
* Checkbox now changes size depending on the card t-shirt size, and the space from the edge of the card to the checkbox also changes based on size/density (aligning with the metadata).
* Updated some spacing/sizing values to match designs.
* Fixed grid definition when there is no action menu.

devongovett commented:
> Only thing is that we might want to hide DnD from the props since isn't ready yet right?

I can do this in a followup
